### ISSUE TITLE
New arrow alias syntax breaks some queries in Sequelize 4.0.0 that worked in 4.0.0-2

### ISSUE THREAD

**Issue by natesilva** [sequelize/sequelize#7742]
When upgrading we found some broken queries. The error message looks like this:

> SequelizeDatabaseError: Unknown column 'purchaseOrder->customer.id' in 'on clause'

(The full query and error message are below.)

Looking at the generated SQL, it creates a `JOIN` using a table alias `purchaseOrder->customer` that is defined in a subquery. That table alias is not in scope in the place where it’s used.

In beta 4.0.0-2 it would have used `purchaseOrder.customer.id`, which works because that is an alias for one of the *column* names exposed by the subquery, and is in scope.

I’ve stripped it down to a basic example. This works in 4.0.0-2, but not in 4.0.0:

```javascript
//
// DB setup
//

const Sequelize = require('sequelize');

const sequelize = new Sequelize('bug_report', 'root', 'password', {
  host: 'localhost',
  port: 32769,
  dialect: 'mysql'
});

const Customer = sequelize.define('customer', {
  name: Sequelize.STRING
});

const ShippingAddress = sequelize.define('shippingAddress', {
  address: Sequelize.STRING,
  verified: Sequelize.BOOLEAN
});

const Order = sequelize.define('purchaseOrder', {
  description: Sequelize.TEXT
});

const Shipment = sequelize.define('shipment', {
  trackingNumber: Sequelize.STRING
});

Customer.hasMany(ShippingAddress);
ShippingAddress.belongsTo(Customer);

Customer.hasMany(Order);
Order.belongsTo(Customer);

Shipment.belongsTo(Order);
Order.hasOne(Shipment);

//
// Bug demo
//

async function f() {
  await sequelize.sync();

  const result = await Shipment.findOne({
    include: [
      {
        model: Order,
        required: true,
        include: [
          {
            model: Customer,
            include: [
              {
                model: ShippingAddress,
                where: { verified: true }
              }
            ]
          }
        ]
      }
    ]
  });

  console.log(result);
}

f();
```

And the generated SQL, and error message is:

```
SELECT `shipment`.*, `purchaseOrder->customer->shippingAddresses`.`id` AS `purchaseOrder.customer.shippingAddresses.id`, `purchaseOrder->customer->shippingAddresses`.`address` AS `purchaseOrder.customer.shippingAddresses.address`, `purchaseOrder->customer->shippingAddresses`.`verified` AS `purchaseOrder.customer.shippingAddresses.verified`, `purchaseOrder->customer->shippingAddresses`.`createdAt` AS `purchaseOrder.customer.shippingAddresses.createdAt`, `purchaseOrder->customer->shippingAddresses`.`updatedAt` AS `purchaseOrder.customer.shippingAddresses.updatedAt`, `purchaseOrder->customer->shippingAddresses`.`customerId` AS `purchaseOrder.customer.shippingAddresses.customerId` FROM (SELECT `shipment`.`id`, `shipment`.`trackingNumber`, `shipment`.`createdAt`, `shipment`.`updatedAt`, `shipment`.`purchaseOrderId`, `purchaseOrder`.`id` AS `purchaseOrder.id`, `purchaseOrder`.`description` AS `purchaseOrder.description`, `purchaseOrder`.`createdAt` AS `purchaseOrder.createdAt`, `purchaseOrder`.`updatedAt` AS `purchaseOrder.updatedAt`, `purchaseOrder`.`customerId` AS `purchaseOrder.customerId`, `purchaseOrder->customer`.`id` AS `purchaseOrder.customer.id`, `purchaseOrder->customer`.`name` AS `purchaseOrder.customer.name`, `purchaseOrder->customer`.`createdAt` AS `purchaseOrder.customer.createdAt`, `purchaseOrder->customer`.`updatedAt` AS `purchaseOrder.customer.updatedAt` FROM `shipments` AS `shipment` INNER JOIN `purchaseOrders` AS `purchaseOrder` ON `shipment`.`purchaseOrderId` = `purchaseOrder`.`id` LEFT OUTER JOIN `customers` AS `purchaseOrder->customer` ON `purchaseOrder`.`customerId` = `purchaseOrder->customer`.`id` LIMIT 1) AS `shipment` INNER JOIN `shippingAddresses` AS `purchaseOrder->customer->shippingAddresses` ON `purchaseOrder->customer.id` = `purchaseOrder->customer->shippingAddresses`.`customerId` AND `purchaseOrder->customer->shippingAddresses`.`verified` = true;
(node:64954) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): SequelizeDatabaseError: Unknown column 'purchaseOrder->customer.id' in 'on clause'
```

__Dialect:__ mysql
__Database version:__ 5.6, 5.7
__Sequelize version:__ 4.0.0

**Comment by sushantdhiman** [comment]
`required: true` is the issue

**Comment by natesilva** [comment]
I guess it doesn’t make sense to have a `Shipment` without an `Order` (!), but this is just a simplified version of some complex models (that weren’t about shipments and orders). The developer wanted to force an `INNER JOIN`.

**Comment by ggrimbert** [comment]
You have the same issue I have. https://github.com/sequelize/sequelize/issues/7710
Can you try the solution I explained ? 
If it works I'll make a PR.

**Comment by natesilva** [comment]
@ggrimbert Yes, I tested a couple of our queries and your fixed worked.

**Comment by natesilva** [comment]
Still an issue, waiting on resolution of #7795, thanks for removing the stale label @sushantdhiman.

**Comment by sushantdhiman** [comment]
https://github.com/sequelize/sequelize/pull/7795

**Comment by natesilva** [comment]
Thank you!
